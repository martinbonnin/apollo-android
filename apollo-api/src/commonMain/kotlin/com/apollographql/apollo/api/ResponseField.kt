package com.apollographql.apollo.api

import kotlin.jvm.JvmStatic

/**
 * An abstraction for a field in a graphQL operation. Field can refer to:
 * - GraphQL
 * - Scalar Types,
 * - Objects
 * - List,
 * - etc.
 *
 * For a complete list of types that a Field object can refer to see [ResponseField.Type] class.
 */
class ResponseField(
    val type: Type,
    val responseName: String,
    val fieldName: String,
    val arguments: Map<String, Any?>,
    val conditions: List<Condition>,
    val possibleFieldSets: Map<String, Array<ResponseField>>,
) {

  /**
   * Resolves field argument value by [name]. If argument represents a references to the variable, it will be resolved from
   * provided operation [variables] values.
   */
  @Suppress("UNCHECKED_CAST")
  fun resolveArgument(
      name: String,
      variables: Operation.Variables
  ): Any? {
    val variableValues = variables.valueMap()
    val argumentValue = arguments[name]
    return if (argumentValue is Map<*, *>) {
      val argumentValueMap = argumentValue as Map<String, Any?>
      if (isArgumentValueVariableType(argumentValueMap)) {
        val variableName = argumentValueMap[VARIABLE_NAME_KEY].toString()
        variableValues[variableName]
      } else {
        null
      }
    } else {
      argumentValue
    }
  }

  sealed class Type {
    class NotNull(val ofType: Type): Type()
    class List(val ofType: Type): Type()

    /**
     * a Named GraphQL type
     *
     * We make the distinction between objects and non-objects ones for the CacheKeyResolver API.
     * In a typical server scenario, the resolvers would have access to the schema and would look up the complete type
     * but we want to stay lightweight so for now we add this information
     */
    sealed class Named(val name: String): Type() {
      class Object(name: String): Named(name)
      class Other(name: String): Named(name)
    }
  }

  enum class Kind { OBJECT, OTHER }
  /**
   * Abstraction for condition to be associated with field
   */
  open class Condition internal constructor() {
    companion object {
      /**
       * Creates new [BooleanCondition] for provided [variableName].
       */
      @JvmStatic
      fun booleanCondition(variableName: String, inverted: Boolean): BooleanCondition {
        return BooleanCondition(variableName, inverted)
      }
    }
  }

  /**
   * Abstraction for boolean condition
   */
  data class BooleanCondition internal constructor(
      val variableName: String,
      val isInverted: Boolean
  ) : Condition()

  companion object {
    private const val VARIABLE_IDENTIFIER_KEY = "kind"
    private const val VARIABLE_IDENTIFIER_VALUE = "Variable"
    const val VARIABLE_NAME_KEY = "variableName"

    @JvmStatic
    fun isArgumentValueVariableType(objectMap: Map<String, Any?>): Boolean {
      return (objectMap.containsKey(VARIABLE_IDENTIFIER_KEY)
          && objectMap[VARIABLE_IDENTIFIER_KEY] == VARIABLE_IDENTIFIER_VALUE && objectMap.containsKey(VARIABLE_NAME_KEY))
    }

    fun Type.leafType(): String = when(this) {
      is Type.NotNull -> ofType.leafType()
      is Type.Named -> name
      is Type.List -> ofType.leafType()
    }

    val Typename = ResponseField(
        type = Type.NotNull(Type.Named.Other("String")),
        responseName = "__typename",
        fieldName = "__typename",
        arguments = emptyMap(),
        conditions = emptyList(),
        possibleFieldSets = emptyMap(),
    )
  }
}
