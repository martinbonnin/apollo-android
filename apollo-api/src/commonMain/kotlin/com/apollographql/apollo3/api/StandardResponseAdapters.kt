package com.apollographql.apollo3.api


import com.apollographql.apollo3.api.internal.json.MapJsonReader.Companion.buffer
import com.apollographql.apollo3.api.internal.json.MapJsonWriter
import com.apollographql.apollo3.api.json.JsonReader
import com.apollographql.apollo3.api.json.JsonWriter
import com.apollographql.apollo3.api.internal.json.Utils
import com.apollographql.apollo3.api.internal.json.Utils.readRecursively

/**
 * This file contains a list of [ResponseAdapter] for standard types
 *
 * They are mostly used from the generated code but could be useful in any other situations that requires adapting from
 * GraphQL to Kotlin.
 * In particular, [AnyResponseAdapter] can be used to read/write a Kotlin representation from/to Json.
 */
class ListResponseAdapter<T>(private val wrappedAdapter: ResponseAdapter<T>) : ResponseAdapter<List<T>> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): List<T> {
    reader.beginArray()
    val list = mutableListOf<T>()
    while (reader.hasNext()) {
      list.add(wrappedAdapter.fromResponse(reader, responseAdapterCache))
    }
    reader.endArray()
    return list
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: List<T>) {
    writer.beginArray()
    value.forEach {
      wrappedAdapter.toResponse(writer, responseAdapterCache, it)
    }
    writer.endArray()
  }
}

class NullableResponseAdapter<T : Any>(private val wrappedAdapter: ResponseAdapter<T>) : ResponseAdapter<T?> {
  init {
    check(wrappedAdapter !is NullableResponseAdapter<*>) {
      "The adapter is already nullable"
    }
  }

  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): T? {
    return if (reader.peek() == JsonReader.Token.NULL) {
      reader.skipValue()
      null
    } else {
      wrappedAdapter.fromResponse(reader, responseAdapterCache)
    }
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: T?) {
    if (value == null) {
      writer.nullValue()
    } else {
      wrappedAdapter.toResponse(writer, responseAdapterCache, value)
    }
  }
}

object StringResponseAdapter : ResponseAdapter<String> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): String {
    return reader.nextString()!!
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: String) {
    writer.value(value)
  }
}

object IntResponseAdapter : ResponseAdapter<Int> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): Int {
    return reader.nextInt()
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: Int) {
    writer.value(value)
  }
}

object DoubleResponseAdapter : ResponseAdapter<Double> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): Double {
    return reader.nextDouble()
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: Double) {
    writer.value(value)
  }
}

object BooleanResponseAdapter : ResponseAdapter<Boolean> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): Boolean {
    return reader.nextBoolean()
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: Boolean) {
    writer.value(value)
  }
}

object AnyResponseAdapter : ResponseAdapter<Any?> {
  fun fromResponse(reader: JsonReader): Any? {
    return reader.readRecursively()
  }

  fun toResponse(writer: JsonWriter, value: Any?) {
    Utils.writeToJson(value, writer)
  }

  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): Any? {
    return reader.readRecursively()
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: Any?) {
    Utils.writeToJson(value, writer)
  }
}

object UploadResponseAdapter : ResponseAdapter<Upload> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): Upload {
    error("File Upload used in output position")
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: Upload) {
    writer.value(value)
  }
}

class ObjectResponseAdapter<T>(
    private val wrappedAdapter: ResponseAdapter<T>,
    private val buffered: Boolean
) : ResponseAdapter<T> {
  override fun fromResponse(reader: JsonReader, responseAdapterCache: ResponseAdapterCache): T {
    val actualReader = if (buffered) {
      reader.buffer()
    } else {
      reader
    }
    actualReader.beginObject()
    return wrappedAdapter.fromResponse(actualReader, responseAdapterCache).also {
      actualReader.endObject()
    }
  }

  override fun toResponse(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache, value: T) {
    if (buffered && writer !is MapJsonWriter) {
      /**
       * Convert to a Map first
       */
      val mapWriter = MapJsonWriter()
      mapWriter.beginObject()
      wrappedAdapter.toResponse(mapWriter, responseAdapterCache, value)
      mapWriter.endObject()

      /**
       * And write to the original writer
       */
      AnyResponseAdapter.toResponse(writer, mapWriter.root())
    } else {
      writer.beginObject()
      wrappedAdapter.toResponse(writer, responseAdapterCache, value)
      writer.endObject()
    }
  }
}

fun <T : Any> ResponseAdapter<T>.nullable() = NullableResponseAdapter(this)
fun <T> ResponseAdapter<T>.list() = ListResponseAdapter(this)
fun <T> ResponseAdapter<T>.obj(buffered: Boolean = false) = ObjectResponseAdapter(this, buffered)

/**
 * Global instances of nullable adapters for built-in scalar types
 */
val NullableStringResponseAdapter = StringResponseAdapter.nullable()
val NullableDoubleResponseAdapter = DoubleResponseAdapter.nullable()
val NullableIntResponseAdapter = IntResponseAdapter.nullable()
val NullableBooleanResponseAdapter = BooleanResponseAdapter.nullable()
