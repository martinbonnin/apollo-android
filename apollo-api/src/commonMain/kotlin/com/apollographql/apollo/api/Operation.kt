package com.apollographql.apollo.api

import com.apollographql.apollo.api.internal.InputFieldMarshaller
import com.apollographql.apollo.api.internal.InputFieldWriter
import com.apollographql.apollo.api.internal.ResponseAdapter
import com.apollographql.apollo.api.internal.Throws
import com.apollographql.apollo.api.internal.json.InputFieldJsonWriter
import com.apollographql.apollo.api.internal.json.JsonWriter
import com.apollographql.apollo.api.internal.json.use
import okio.Buffer
import okio.IOException
import kotlin.jvm.JvmField

/**
 * Represents a GraphQL operation (mutation or query).
 */
interface Operation<D : Operation.Data> {
  /**
   * Returns the raw GraphQL operation String.
   */
  fun queryDocument(): String

  /**
   * Returns the variables associated with this GraphQL operation.
   */
  fun variables(): Variables

  /**
   * Returns an Adapter that maps the server response data to/from generated model class [D].
   */
  fun adapter(): ResponseAdapter<D>

  /**
   * Returns the array of root fields
   */
  fun responseFields(): Array<ResponseField>

  /**
   * Returns GraphQL operation name
   */
  fun name(): String

  /**
   * Returns a unique identifier for this operation.
   */
  fun operationId(): String

  /**
   * Marker interface for generated models built from data returned by the server in response to this operation.
   */
  interface Data

  /**
   * Abstraction for the variables which are a part of the GraphQL operation.
   * For example, for the following GraphQL operation, Variables represents values for GraphQL '$type' and '$limit' variables:
   *
   * ```
   *  query FeedQuery($type: FeedType!, $limit: Int!) {
   *    feedEntries: feed(type: $type, limit: $limit) {
   *      id
   *      repository {
   *        ...RepositoryFragment
   *      }
   *      postedBy {
   *        login
   *      }
   *    }
   * }
   * ```
   */
  open class Variables {

    open fun valueMap(): Map<String, Any?> {
      return emptyMap()
    }

    open fun marshaller(): InputFieldMarshaller {
      return object : InputFieldMarshaller {
        override fun marshal(writer: InputFieldWriter) {
          // noop
        }
      }
    }

    /**
     * Serializes variables as JSON string to be sent to the GraphQL server.
     */
    @Throws(IOException::class)
    fun marshal(): String {
      return marshal(CustomScalarAdapters.DEFAULT)
    }

    /**
     * Serializes variables with provided scalarTypeAdapters [customScalarAdapters] as JSON string to be sent to the GraphQL server.
     */
    @Throws(IOException::class)
    fun marshal(customScalarAdapters: CustomScalarAdapters): String {
      return Buffer().apply {
        JsonWriter.of(this).use { jsonWriter ->
          jsonWriter.serializeNulls = true
          jsonWriter.beginObject()
          marshaller().marshal(InputFieldJsonWriter(jsonWriter, customScalarAdapters))
          jsonWriter.endObject()
        }
      }.readUtf8()
    }
  }

  companion object {

    @JvmField
    val EMPTY_VARIABLES = Variables()
  }
}
