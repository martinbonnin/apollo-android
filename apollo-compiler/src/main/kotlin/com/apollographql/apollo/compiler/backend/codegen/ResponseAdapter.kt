package com.apollographql.apollo.compiler.backend.codegen

import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.internal.ResponseAdapter
import com.apollographql.apollo.compiler.applyIf
import com.apollographql.apollo.compiler.backend.ast.CodeGenerationAst
import com.apollographql.apollo.compiler.escapeKotlinReservedWord
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.asTypeName
import com.squareup.kotlinpoet.joinToCode

internal fun CodeGenerationAst.OperationType.responseAdapterTypeSpec(generateAsInternal: Boolean = false): TypeSpec {
  return this.dataType
      .copy(name = "${this.name.escapeKotlinReservedWord()}_ResponseAdapter")
      .rootResponseAdapterTypeSpec(generateAsInternal)
}

internal fun CodeGenerationAst.FragmentType.responseAdapterTypeSpec(generateAsInternal: Boolean = false): TypeSpec {
  val dataType = this.implementationType.nestedObjects.single()
  return dataType
      .copy(name = "${this.implementationType.name.escapeKotlinReservedWord()}_ResponseAdapter")
      .rootResponseAdapterTypeSpec(generateAsInternal)
}

private fun CodeGenerationAst.ObjectType.rootResponseAdapterTypeSpec(generateAsInternal: Boolean = false): TypeSpec {
  return this.responseAdapterTypeSpec()
      .toBuilder()
      .applyIf(generateAsInternal) { addModifiers(KModifier.INTERNAL) }
      .addAnnotation(suppressWarningsAnnotation)
      .build()
}

private fun CodeGenerationAst.ObjectType.responseAdapterTypeSpec(): TypeSpec {
  return TypeSpec.objectBuilder(this.name)
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(this.typeRef.asTypeName()))
      .applyIf(
          this.fields.isNotEmpty()
              && this.kind is CodeGenerationAst.ObjectType.Kind.Object) {
        addProperty(responseFieldsPropertySpec(this@responseAdapterTypeSpec))
      }
      .addFunction(readFromResponseFunSpec())
      .addFunction(writeToResponseFunSpec())
      .addTypes(
          this.nestedObjects
              .filter {
                it.kind is CodeGenerationAst.ObjectType.Kind.Object
                    || (it.kind is CodeGenerationAst.ObjectType.Kind.Fragment && it.kind.possibleImplementations.isNotEmpty())
              }
              .map { nestedObject ->
                nestedObject.responseAdapterTypeSpec()
              }
      )
      .build()
}

internal fun CodeGenerationAst.TypeRef.asAdapterTypeName(): ClassName {

  return when {
    // We are called on a root typeRef such as TestQuery, just add '_ResponseAdapter'
    enclosingType == null -> {
      ClassName(packageName = "$packageName.adapter", "${this.name.escapeKotlinReservedWord()}_ResponseAdapter")
    }
    // We are called on a data typeRef such as TestQuery.Data, skip Data
    enclosingType.enclosingType == null -> {
      ClassName(packageName = "$packageName.adapter", "${this.enclosingType.name.escapeKotlinReservedWord()}_ResponseAdapter")
    }
    else -> {
      ClassName(packageName = "$packageName.adapter", enclosingType.asAdapterTypeName().simpleNames + this.name.escapeKotlinReservedWord())
    }
  }
}


private fun responseFieldsPropertySpec(objectType: CodeGenerationAst.ObjectType): PropertySpec {
  val fields = objectType.fields

  val initializer = CodeBlock.builder()
      .addStatement("arrayOf(")
      .indent()
      .also { builder->
        builder.add(fields.map { field -> field.responseFieldInitializerCode(objectType) }.joinToCode(separator = ",\n"))
      }
      .unindent()
      .addStatement("")
      .add(")")
      .build()
  return PropertySpec
      .builder(
          name = "RESPONSE_FIELDS",
          type = Array<ResponseField>::class.asClassName().parameterizedBy(ResponseField::class.asClassName()),
      )
      .initializer(initializer)
      .build()
}

private fun CodeGenerationAst.FieldType.toTypeCode(): CodeBlock {
  val builder = CodeBlock.Builder()

  when (this) {
    is CodeGenerationAst.FieldType.Object -> {
      builder.add("%T(%S)", ResponseField.Type.Named.Object::class.java, schemaTypeName)
    }
    is CodeGenerationAst.FieldType.Scalar -> {
      // same code as Object but in a separate branch so that type inference can find schemaTypeName
      builder.add("%T(%S)", ResponseField.Type.Named.Other::class.java, schemaTypeName)
    }
    is CodeGenerationAst.FieldType.Array -> {
      builder.add("%T(", ResponseField.Type.List::class.java)
      builder.add(this.rawType.toTypeCode())
      builder.add(")")
    }
  }

  val block = builder.build()

  return if (!nullable) {
    block.toNonNullable()
  } else
    block
}

private fun CodeBlock.toNonNullable(): CodeBlock {
  val builder =  CodeBlock.builder()
  builder.add("%T(", ResponseField.Type.NotNull::class.java)
  builder.add(this)
  builder.add(")")
  return builder.build()
}

private fun CodeGenerationAst.Field.responseFieldInitializerCode(objectType: CodeGenerationAst.ObjectType): CodeBlock {
    val builder = CodeBlock.builder().add("%T(\n", ResponseField::class)
    builder.indent()
    builder.add("type = %L,\n", type.toTypeCode())
    builder.add("responseName = %S,\n", responseName)
    builder.add("fieldName = %S,\n", schemaName)
    builder.add("arguments = %L,\n", arguments.takeIf { it.isNotEmpty() }?.let { anyToCode(it) } ?: "emptyMap()")
    builder.add("conditions = %L,\n", conditionsListCode(conditions))
    builder.add("possibleFieldSets = %L,\n", fieldsCode(this.type, objectType))
    builder.unindent()
    builder.add(")")

    return builder.build()
  }

private fun fieldsCode(type: CodeGenerationAst.FieldType, objectType: CodeGenerationAst.ObjectType): CodeBlock {
  return when (val leafType = type.leafType()) {
    is CodeGenerationAst.FieldType.Scalar -> CodeBlock.of("emptyMap()")
    is CodeGenerationAst.FieldType.Object -> {
      val nestedObjectType = objectType.nestedObjects.first { it.typeRef == leafType.typeRef }
      val builder = CodeBlock.Builder()
      builder.add("mapOf(\n")
      builder.indent()
      when (val kind = nestedObjectType.kind) {
        is CodeGenerationAst.ObjectType.Kind.Object -> {
          builder.add("%S to %T.RESPONSE_FIELDS\n", "", leafType.typeRef.asAdapterTypeName())
        }
        is CodeGenerationAst.ObjectType.Kind.Fragment -> {
          kind.possibleImplementations.forEach {
            builder.add("%S to %T.RESPONSE_FIELDS,\n", it.key, it.value.asAdapterTypeName())
          }
          builder.add("%S to %T.RESPONSE_FIELDS,\n", "", kind.defaultImplementation.asAdapterTypeName())
        }
      }
      builder.unindent()
      builder.add(")")
      builder.build()
    }
    else -> error("")
  }
}

private fun CodeGenerationAst.FieldType.leafType(): CodeGenerationAst.FieldType = when(this) {
  is CodeGenerationAst.FieldType.Scalar -> this
  is CodeGenerationAst.FieldType.Object -> this
  is CodeGenerationAst.FieldType.Array -> rawType.leafType()
}
private fun conditionsListCode(conditions: Set<CodeGenerationAst.Field.Condition>): CodeBlock {
  return conditions
      .map { condition ->
        when (condition) {
          is CodeGenerationAst.Field.Condition.Directive -> CodeBlock.of("%T.booleanCondition(%S,·%L)",
              ResponseField.Condition::class, condition.variableName.escapeKotlinReservedWord(), condition.inverted)
        }
      }
      .joinToCode(separator = ",\n")
      .let {
        if (conditions.isEmpty()) {
          CodeBlock.of("emptyList()")
        } else {
          CodeBlock.builder()
              .add("listOf(\n")
              .indent().add(it).unindent()
              .add("\n)")
              .build()
        }
      }
}

private fun anyToCode(any: Any?): CodeBlock {
  return with(any) {
    when {
      this == null -> CodeBlock.of("null")
      this is Map<*, *> && this.isEmpty() -> CodeBlock.of("emptyMap<%T,·Any?>()", String::class.asTypeName())
      this is Map<*, *> -> CodeBlock.builder()
          .add("mapOf<%T,·Any?>(\n", String::class.asTypeName())
          .indent()
          .add(map { CodeBlock.of("%S to %L", it.key, anyToCode(it.value)) }.joinToCode(separator = ",\n"))
          .unindent()
          .add(")")
          .build()
      this is List<*> && this.isEmpty() -> CodeBlock.of("emptyList<Any?>()")
      this is List<*> -> CodeBlock.builder()
          .add("listOf<Any?>(\n")
          .indent()
          .add(map { anyToCode(it) }.joinToCode(separator = ",\n"))
          .unindent()
          .add(")")
          .build()
      this is String -> CodeBlock.of("%S", this)
      this is Number -> CodeBlock.of("%L", this)
      this is Boolean -> CodeBlock.of("%L", this)
      else -> throw IllegalStateException("Cannot generate code for $this")
    }
  }
}