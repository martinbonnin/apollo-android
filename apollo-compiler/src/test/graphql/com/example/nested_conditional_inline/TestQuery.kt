// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package com.example.nested_conditional_inline

import com.apollographql.apollo.api.Input
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.Query
import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.internal.InputFieldMarshaller
import com.apollographql.apollo.api.internal.QueryDocumentMinifier
import com.apollographql.apollo.api.internal.ResponseAdapter
import com.example.nested_conditional_inline.adapter.TestQuery_ResponseAdapter
import com.example.nested_conditional_inline.type.Episode
import kotlin.Any
import kotlin.Array
import kotlin.Double
import kotlin.String
import kotlin.Suppress
import kotlin.collections.List
import kotlin.collections.Map
import kotlin.jvm.Transient

@Suppress("NAME_SHADOWING", "UNUSED_ANONYMOUS_PARAMETER", "LocalVariableName",
    "RemoveExplicitTypeArguments", "NestedLambdaShadowedImplicitParameter", "PropertyName",
    "RemoveRedundantQualifierName")
data class TestQuery(
  val episode: Input<Episode> = Input.absent()
) : Query<TestQuery.Data> {
  @Transient
  private val variables: Operation.Variables = object : Operation.Variables() {
    override fun valueMap(): Map<String, Any?> = mutableMapOf<String, Any?>().apply {
      if (this@TestQuery.episode.defined) {
        this["episode"] = this@TestQuery.episode.value
      }
    }

    override fun marshaller(): InputFieldMarshaller {
      return InputFieldMarshaller.invoke { writer ->
        if (this@TestQuery.episode.defined) {
          writer.writeString("episode", this@TestQuery.episode.value?.rawValue)
        }
      }
    }
  }

  override fun operationId(): String = OPERATION_ID

  override fun queryDocument(): String = QUERY_DOCUMENT

  override fun variables(): Operation.Variables = variables

  override fun name(): String = OPERATION_NAME

  override fun adapter(): ResponseAdapter<Data> = TestQuery_ResponseAdapter
  override fun responseFields(): Array<ResponseField> = TestQuery_ResponseAdapter.RESPONSE_FIELDS
  /**
   * The query type, represents all of the entry points into our object graph
   */
  data class Data(
    val hero: Hero?
  ) : Operation.Data {
    /**
     * A character from the Star Wars universe
     */
    interface Hero {
      val __typename: String

      /**
       * The name of the character
       */
      val name: String

      interface Human : Hero {
        override val __typename: String

        /**
         * The name of the character
         */
        override val name: String

        /**
         * This human's friends, or an empty list if they have none
         */
        val friends: List<Friend?>?

        /**
         * A character from the Star Wars universe
         */
        interface Friend {
          val __typename: String

          /**
           * The name of the character
           */
          val name: String

          interface Human : Friend {
            override val __typename: String

            /**
             * The name of the character
             */
            override val name: String

            /**
             * Height in the preferred unit, default is meters
             */
            val height: Double?
          }

          companion object {
            fun Friend.asHuman(): Human? = this as? Human
          }
        }
      }

      interface Droid : Hero {
        override val __typename: String

        /**
         * The name of the character
         */
        override val name: String

        /**
         * This droid's friends, or an empty list if they have none
         */
        val friends: List<Friend?>?

        /**
         * A character from the Star Wars universe
         */
        interface Friend {
          val __typename: String

          /**
           * The name of the character
           */
          val name: String

          interface Human : Friend {
            override val __typename: String

            /**
             * The name of the character
             */
            override val name: String

            /**
             * Height in the preferred unit, default is meters
             */
            val height: Double?
          }

          companion object {
            fun Friend.asHuman(): Human? = this as? Human
          }
        }
      }

      data class HumanHero(
        override val __typename: String,
        /**
         * The name of the character
         */
        override val name: String,
        /**
         * This human's friends, or an empty list if they have none
         */
        override val friends: List<Friend?>?
      ) : Hero, Human {
        /**
         * A character from the Star Wars universe
         */
        interface Friend : Human.Friend {
          override val __typename: String

          data class HumanFriend(
            override val __typename: String,
            /**
             * The name of the character
             */
            override val name: String,
            /**
             * Height in the preferred unit, default is meters
             */
            override val height: Double?
          ) : Human.Friend, Human.Friend.Human, Friend

          data class OtherFriend(
            override val __typename: String,
            /**
             * The name of the character
             */
            override val name: String
          ) : Human.Friend, Friend
        }
      }

      data class DroidHero(
        override val __typename: String,
        /**
         * The name of the character
         */
        override val name: String,
        /**
         * This droid's friends, or an empty list if they have none
         */
        override val friends: List<Friend?>?
      ) : Hero, Droid {
        /**
         * A character from the Star Wars universe
         */
        interface Friend : Droid.Friend {
          override val __typename: String

          data class HumanFriend(
            override val __typename: String,
            /**
             * The name of the character
             */
            override val name: String,
            /**
             * Height in the preferred unit, default is meters
             */
            override val height: Double?
          ) : Droid.Friend, Droid.Friend.Human, Friend

          data class OtherFriend(
            override val __typename: String,
            /**
             * The name of the character
             */
            override val name: String
          ) : Droid.Friend, Friend
        }
      }

      data class OtherHero(
        override val __typename: String,
        /**
         * The name of the character
         */
        override val name: String
      ) : Hero

      companion object {
        fun Hero.asHuman(): Human? = this as? Human

        fun Hero.asDroid(): Droid? = this as? Droid
      }
    }
  }

  companion object {
    const val OPERATION_ID: String =
        "a9f066a7d1092096ab154f16f32114a4bd71e959b789f37879249cdf6309ea86"

    val QUERY_DOCUMENT: String = QueryDocumentMinifier.minify(
          """
          |query TestQuery(${'$'}episode: Episode) {
          |  hero(episode: ${'$'}episode) {
          |    __typename
          |    name
          |    ... on Human {
          |      friends {
          |        __typename
          |        name
          |        ... on Human {
          |          height(unit: FOOT)
          |        }
          |      }
          |    }
          |    ... on Droid {
          |      friends {
          |        __typename
          |        name
          |        ... on Human {
          |          height(unit: METER)
          |        }
          |      }
          |    }
          |  }
          |}
          """.trimMargin()
        )

    val OPERATION_NAME: String = "TestQuery"
  }
}
