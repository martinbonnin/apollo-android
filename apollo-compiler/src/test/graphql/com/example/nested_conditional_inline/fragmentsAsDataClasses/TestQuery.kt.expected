// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package com.example.nested_conditional_inline

import com.apollographql.apollo3.api.Input
import com.apollographql.apollo3.api.Query
import com.apollographql.apollo3.api.QueryDocumentMinifier
import com.apollographql.apollo3.api.ResponseAdapter
import com.apollographql.apollo3.api.ResponseAdapterCache
import com.apollographql.apollo3.api.ResponseField
import com.apollographql.apollo3.api.json.JsonWriter
import com.apollographql.apollo3.api.obj
import com.example.nested_conditional_inline.adapter.TestQuery_ResponseAdapter
import com.example.nested_conditional_inline.adapter.TestQuery_VariablesAdapter
import com.example.nested_conditional_inline.type.Episode
import kotlin.Double
import kotlin.String
import kotlin.Suppress
import kotlin.collections.List

@Suppress("NAME_SHADOWING", "UNUSED_ANONYMOUS_PARAMETER", "LocalVariableName",
    "RemoveExplicitTypeArguments", "NestedLambdaShadowedImplicitParameter", "PropertyName",
    "RemoveRedundantQualifierName")
data class TestQuery(
  val episode: Input<Episode?> = Input.Absent
) : Query<TestQuery.Data> {
  override fun operationId(): String = OPERATION_ID

  override fun queryDocument(): String = QUERY_DOCUMENT

  override fun serializeVariables(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache) {
    TestQuery_VariablesAdapter.obj().toResponse(writer, responseAdapterCache, this)
  }

  override fun name(): String = OPERATION_NAME

  override fun adapter(): ResponseAdapter<Data> {
    return TestQuery_ResponseAdapter.obj()
  }

  override fun responseFields(): List<ResponseField.FieldSet> = listOf(
    ResponseField.FieldSet(null, TestQuery_ResponseAdapter.RESPONSE_FIELDS)
  )
  /**
   * The query type, represents all of the entry points into our object graph
   */
  data class Data(
    val hero: Hero?
  ) : Query.Data {
    /**
     * A character from the Star Wars universe
     */
    data class Hero(
      val __typename: String = "Character",
      /**
       * The name of the character
       */
      val name: String,
      val asAsHuman: AsHuman?,
      val asAsDroid: AsDroid?
    ) {
      data class AsHuman(
        val __typename: String,
        /**
         * The name of the character
         */
        val name: String,
        /**
         * This human's friends, or an empty list if they have none
         */
        val friends: List<Friends?>?
      ) {
        fun friendsFilterNotNull(): List<Friends>? = friends?.filterNotNull()

        /**
         * A character from the Star Wars universe
         */
        data class Friends(
          val __typename: String = "Character",
          /**
           * The name of the character
           */
          val name: String,
          val asAsHuman: AsHuman?
        ) {
          data class AsHuman(
            val __typename: String,
            /**
             * The name of the character
             */
            val name: String,
            /**
             * Height in the preferred unit, default is meters
             */
            val height: Double?
          )
        }
      }

      data class AsDroid(
        val __typename: String,
        /**
         * The name of the character
         */
        val name: String,
        /**
         * This droid's friends, or an empty list if they have none
         */
        val friends: List<Friends?>?
      ) {
        fun friendsFilterNotNull(): List<Friends>? = friends?.filterNotNull()

        /**
         * A character from the Star Wars universe
         */
        data class Friends(
          val __typename: String = "Character",
          /**
           * The name of the character
           */
          val name: String,
          val asAsHuman: AsHuman?
        ) {
          data class AsHuman(
            val __typename: String,
            /**
             * The name of the character
             */
            val name: String,
            /**
             * Height in the preferred unit, default is meters
             */
            val height: Double?
          )
        }
      }
    }
  }

  companion object {
    const val OPERATION_ID: String =
        "a9f066a7d1092096ab154f16f32114a4bd71e959b789f37879249cdf6309ea86"

    val QUERY_DOCUMENT: String = QueryDocumentMinifier.minify(
          """
          |query TestQuery(${'$'}episode: Episode) {
          |  hero(episode: ${'$'}episode) {
          |    __typename
          |    name
          |    ... on Human {
          |      friends {
          |        __typename
          |        name
          |        ... on Human {
          |          height(unit: FOOT)
          |        }
          |      }
          |    }
          |    ... on Droid {
          |      friends {
          |        __typename
          |        name
          |        ... on Human {
          |          height(unit: METER)
          |        }
          |      }
          |    }
          |  }
          |}
          """.trimMargin()
        )

    val OPERATION_NAME: String = "TestQuery"
  }
}
